### ISSUE TITLE
Before/Finally should be scoped to :all and :each

### ISSUE THREAD

**Issue by erikreedstrom** [antonmi/espec#73]
There are many cases where code should run only once for all examples within a group. For instance, using ExMachina to generate from a factory within a before block will fail if there are any unique constraints, as it attempts to save on each example.

Rspec provides `before(:all)` and `after(:all)` to complement the `:each`, yet it does not seem this is available in espec currently.

**Comment by antonmi** [comment]
Hi, Erik!
My experience shows that RSpec's before(:all) causes more problems than it solves.
It creates a 'state' in test cases. That state can be changed in one specs and, therefore, affect the others.
I can point three reasons why I don't want such feature in ESpec.
1. 'Testing best practices' teach us to avoid shared states.
2. ESpec specs are executed in random order by default, so presence of before(:all) can cause a lot of confusion.
3. Espec has asynchronous test execution feature (async: true). It's difficult to figure out how before(:all) block should be executed (particularly in cases when there are several before(:all) block in different contexts).
But I'm open for discussion.
Thanks!

**Comment by dm1try** [comment]
>It creates a 'state' in test cases. That state can be changed in one specs and, therefore, affect the others.
I can point three reasons why I don't want such feature in ESpec.
1. 'Testing best practices' teach us to avoid shared states.

The good point. But in some cases I want to share the initial state; ex., it helps to skip the heavy initialization in high-level integration tests:

```elixir
  defmodule EmbedNvim do
    def ensure_started do
      if Process.whereis(__MODULE__) == nil do
        Session.attach(
          %ErlangPort.Address{ param: [{:spawn, "nvim --embed" }, [ env: [ { 'XDG_CONFIG_HOME', System.get_env("NVIM_CONFIG_HOME") } ]]]},
          [session_name: __MODULE__]
        )
      end
    end

    def command(command) do
      {:ok, response} = Session.call(__MODULE__, "vim_command_output", [command])
      String.strip(response)
    end
  end

  before do
    EmbedNvim.ensure_started
  end

  it "registers the test plugin" do
      response = EmbedNvim.command("UpdateRemotePlugins")
      expect(response).to match "elixir host registered plugins ['test_plugin.exs']"
    end
  end
```
How it works now :) In such case I can do a partial cleanup(closing all open files in vim instance for example) in `before/after` blocks but I don't need to cleanup the initial state.

I think we need analog of `setup_all` from `ExUnit`(?). We can deny using `before(:all)/after(:all)` in nested  contexts :grimacing:

**Comment by erikreedstrom** [comment]
I will concede that for my purposes I can recreate a given state for each test, but this requires that I can effectively tear down state after a failure. Currently this is not possible with #75

**Comment by jejacks0n** [comment]
Came here to resolve something similar, and while the information in here does capture what I need, I figured it'd be worth adding my thoughts as well. I've actually never used a before(:all) for managing some state, because that seems generally bad. But, I have absolutely used it to setup something and tear something down -- for instance a pubsub gets setup before all tests, and the pubsub is shutdown at the end (which may do some cleanup managed by essentially a manager). A good example of this might be something tying in to RabbitMQ and that needs to be available for the tests.

So basically I'm saying that maybe a before(:all) isn't needed, but an after(:all) is.. and so if the after(:all) concept is added, might as well get closer to parity. =)

**Comment by antonmi** [comment]
@erikreedstrom , @dm1try , @jejacks0n !
Sorry for long inactivity.
Thanks for the discussion!
I think it is a good idea to add smth like setup_all to ESpec. Just one block per spec module.
Will start working on it soon!

**Comment by antonmi** [comment]
@erikreedstrom @jejacks0n @dm1try 
Meet the hooks in 1.2.0!
Thanks for your ideas!
